### PR TITLE
[example] mqtt example working with Google Cloud sample server

### PR DESCRIPTION
**Pull request by gjc13**
This PR includes the commits from travis PR to enable travis.
Corresponding changes can be ignored.

### CODE REVIEW

**Comment by bukepo** [comment]
rebase to master?

File: examples/apps/test/mqtt.cpp [hunk 1]
```diff
@@ -64,21 +68,22 @@ static const unsigned long kMaxConnectIntervalMillis         = 6000L;
...
-    if (aResult == ERR_OK)
-    {
-        printf("Publish done\r\n");
-    }
-    else
-    {
-        printf("Publish get error %d\r\n", aResult);
-    }
+    snprintf(aIat, (size_t)time_size, "%zu", (size_t)now_seconds);
+    snprintf(aExt, (size_t)time_size, "%zu", (size_t)(now_seconds + 3600));
```
Comment (bukepo) [inline examples/apps/test/mqtt.cpp]: Use static_cast

File: examples/apps/test/mqtt.cpp [hunk 2]
```diff
@@ -64,21 +68,22 @@ static const unsigned long kMaxConnectIntervalMillis         = 6000L;
...
 
-void GoogleCloudIotMqttClient::mqttConnectChanged(mqtt_client_t *aClient, void *aArg, mqtt_connection_status_t aResult)
+void GoogleCloudIotMqttClient::sMqttPubSubChanged(void *aArg, err_t aResult)
+{
+    ConnectContext *ctx         = static_cast<ConnectContext *>(aArg);
+    ctx->mClient->mPubSubResult = aResult;
+    xTaskNotify(ctx->mHandle, MQTT_PUBSUB_NOTIFY_VALUE, eSetBits);
+}
+
+void GoogleCloudIotMqttClient::sMqttConnectChanged(mqtt_client_t *aClient, void *aArg, mqtt_connection_status_t aResult)
```
Comment (bukepo) [inline examples/apps/test/mqtt.cpp]: No prefix `s`

File: examples/apps/test/mqtt.cpp [hunk 3]
```diff
@@ -181,7 +179,7 @@ int GoogleCloudIotMqttClient::Connect()
         serverAddr.type = IPADDR_TYPE_V6;
 
         LOCK_TCPIP_CORE();
-        mqtt_client_connect(mMqttClient, &serverAddr, 8883, &GoogleCloudIotMqttClient::mqttConnectChanged, &ctx,
+        mqtt_client_connect(mMqttClient, &serverAddr, 8883, &GoogleCloudIotMqttClient::sMqttConnectChanged, &ctx,
```
Comment (bukepo) [inline examples/apps/test/mqtt.cpp]: Declare this magic number?

File: examples/apps/test/mqtt.cpp [hunk 4]
```diff
@@ -199,14 +197,89 @@ int GoogleCloudIotMqttClient::Connect()
...
+        mSubDataBuf[mDataOffset] = 0;
+        if (mSubCb)
+        {
+            mSubCb(mSubTopicNameBuf, mSubDataBuf, mDataOffset);
+        }
+        mDataOffset = 0;
+    }
+}
+
+void GoogleCloudIotMqttClient::sMqttPublishCallback(void *aArg, const char *aTopic, uint32_t aTotalLength)
```
Comment (bukepo) [inline examples/apps/test/mqtt.cpp]: why totalLength?

File: examples/apps/test/user.cpp [hunk 5]
```diff
@@ -61,7 +71,51 @@ static void ProcessTest(int argc, char *argv[])
...
+            strncpy(sProject, argv[3], sizeof(sProject));
+            strncpy(sRegion, argv[4], sizeof(sRegion));
+            snprintf(sClientId, sizeof(sClientId), "projects/%s/locations/%s/registries/%s/devices/%s", sProject,
+                     sRegion, sRegistry, sDevice);
+            {
+                FILE *fp = fopen(argv[5], "r");
+
+                if (fp)
+                {
+                    size_t n = fread(sPrivKey, 1, sizeof(sPrivKey), fp);
```
Comment (bukepo) [inline examples/apps/test/user.cpp]: dynamic generate .h file so that all platform are supported?

File: examples/apps/test/mqtt.cpp [hunk 6]
```diff
@@ -225,40 +298,42 @@ GoogleCloudIotMqttClient::~GoogleCloudIotMqttClient()
 } // namespace app
 } // namespace ot
 
-void mqttTask(void *p)
+void configCallback(const char *aTopic, const char *aMsg, uint16_t aMsgLength)
```
Comment (bukepo) [inline examples/apps/test/mqtt.cpp]: declared as `static`?

File: examples/apps/test/mqtt.cpp [hunk 7]
```diff
@@ -225,40 +298,42 @@ GoogleCloudIotMqttClient::~GoogleCloudIotMqttClient()
...
-    cfg.mRegistryId      = CLOUDIOT_REGISTRY_ID;
-    cfg.mProjectId       = CLOUDIOT_PROJECT_ID;
-    cfg.mRegion          = CLOUDIOT_REGION;
-    cfg.mRootCertificate = CLOUDIOT_CERT;
-    cfg.mPrivKey         = CLOUDIOT_PRIV_KEY;
-    cfg.mAlgorithm       = JWT_ALG_RS256;
+void mqttTask(void *p)
+{
+    ot::app::GoogleCloudIotClientCfg *cfg  = static_cast<ot::app::GoogleCloudIotClientCfg *>(p);
+    char                              subTopic[50];
```
Comment (bukepo) [inline examples/apps/test/mqtt.cpp]: avoid magic number?

File: include/apps/google_cloud_iot/mqtt_client.hpp [hunk 8]
```diff
@@ -54,24 +54,42 @@ struct GoogleCloudIotClientCfg
...
 
     struct mqtt_connect_client_info_t mClientInfo;
     mqtt_client_t *                   mMqttClient;
     mqtt_connection_status_t          mConnectResult;
+    int                               mPubSubResult;
+
+    MqttTopicDataCallback mSubCb;
+
+    // Currently we only support short message with small buffers
+    char     mSubTopicNameBuf[50];
```
Comment (bukepo) [inline include/apps/google_cloud_iot/mqtt_client.hpp]: avoid magic numbers?

File: examples/apps/test/mqtt.cpp (hunk 4 above)
Comment (gjc13) [reply examples/apps/test/mqtt.cpp]: This argument is not the length of the topic but the length of the whole mqtt message.

File: examples/apps/include/google_cloud_iot/client_cfg.h [hunk 9]
```diff
@@ -0,0 +1,105 @@
+/*
```
Comment (bukepo) [inline examples/apps/include/google_cloud_iot/client_cfg.h]: do we need this?

**Comment by gjc13** [comment]
@bukepo Friendly reminder

**Review by bukepo — APPROVED** [review]
Let's merge this, and enhance style later.

File: examples/apps/test/user.cpp [hunk 10]
```diff
@@ -35,7 +35,11 @@
 #include <openthread/error.h>
 #include <openthread/openthread-freertos.h>
 
-TaskHandle_t gTestTask = NULL;
+#include "google_cloud_iot/client_cfg.h"
+#include "google_cloud_iot/mqtt_client.hpp"
+
+TaskHandle_t                     gTestTask = NULL;
+ot::app::GoogleCloudIotClientCfg sCloudIotCfg;
```
Comment (bukepo) [inline examples/apps/test/user.cpp]: add `static`?

File: examples/apps/include/google_cloud_iot/mqtt_client.hpp [hunk 11]
```diff
@@ -54,24 +54,46 @@ struct GoogleCloudIotClientCfg
...
+
+    int Publish(const char *aTopic, const char *aMsg, size_t aMsgLength);
 
-    int Publish(const char *aTopi, const uint8_t *aMsg, size_t aMsgLength);
+    int Subscribe(const char *aTopic, MqttTopicDataCallback aCb);
 
-    ~GoogleCloudIotMqttClient();
+    ~GoogleCloudIotMqttClient(void);
+
+    static const int kTopicNameMaxLength = 50;
```
Comment (bukepo) [inline examples/apps/include/google_cloud_iot/mqtt_client.hpp]: use `const size_t`?

**Comment by gjc13** [comment]
@jwhui Friendly reminder

**Review by jwhui — APPROVED** [review]

**Pull request merged**